File: ios/Tests/AksaraTests/TestSupport.swift
import Foundation
@testable import Aksara

enum TestSupport {
    /// A throwaway directory unique per call — keeps each test's disk cache isolated.
    static func makeTempDir() -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("AksaraTests-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }

    static func data(_ json: String) -> Data { Data(json.utf8) }
}

/// A custom parser for a non-i18next format: `{"items":[{"id":"a.b","text":"…"}]}`.
/// Demonstrates injecting a consumer-defined JSON model via `LocalizationConfig.parser`.
struct ListParser: TranslationParser {
    struct Doc: Decodable {
        struct Item: Decodable { let id: String; let text: String }
        let items: [Item]
    }

    func parse(_ data: Data, language: String) throws -> [String: String] {
        let doc = try JSONDecoder().decode(Doc.self, from: data)
        return Dictionary(uniqueKeysWithValues: doc.items.map { ($0.id, $0.text) })
    }
}

File: android/aksara-core/src/test/kotlin/com/aksara/TestSupport.kt
package com.aksara

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import kotlin.io.path.createTempDirectory

object TestSupport {
    /** A throwaway directory unique per call — keeps each test's disk cache isolated. */
    fun makeTempDir(): File = createTempDirectory("AksaraTests").toFile()

    /** Loads bundled test JSON (`en.json`, ...) from the test classpath. */
    fun classpathLoader(): (String) -> ByteArray? = { language ->
        TestSupport::class.java.classLoader
            ?.getResourceAsStream("$language.json")
            ?.use { it.readBytes() }
    }

    fun bytes(json: String): ByteArray = json.toByteArray()
}

/**
 * A custom parser for a non-i18next format: `{"items":[{"id":"a.b","text":"…"}]}`.
 * Demonstrates injecting a consumer-defined JSON model via [LocalizationConfig.parser].
 */
class ListParser : TranslationParser {
    override fun parse(data: ByteArray, language: String): Map<String, String> {
        val items = Json.parseToJsonElement(data.decodeToString()).jsonObject["items"]!!.jsonArray
        return items.associate { item ->
            val obj = item.jsonObject
            obj["id"]!!.jsonPrimitive.content to obj["text"]!!.jsonPrimitive.content
        }
    }
}
